app/backends/workflow-orchestration.ts

```diff
@@ -526,138 +537,162 @@ export async function handleWorkItemUpdate(
-        nextStep = await createNextWorkItems(
-          tx,
-          logger,
-          workItem,
-          allWorkItemsForStepComplete,
-          results,
-          outputItemSizes);
+      await updateWorkItemStatus(
+        tx,
+        workItemID,
+        status as WorkItemStatus,
+        duration,
+        totalItemsSize,
+        outputItemSizes);
+
+      const completedWorkItemCount = await workItemCountForStep(
+        tx, workItem.jobID, workItem.workflowStepIndex, COMPLETED_WORK_ITEM_STATUSES,
+      );
+      const allWorkItemsForStepComplete = (completedWorkItemCount == thisStep.workItemCount);
+
+      if (hits && job.numInputGranules > hits) {
+        job.numInputGranules = hits;
+        await job.save(tx);
+        await updateWorkItemCounts(tx, job);
       }
 
-      if (nextStep) {
-        if (results && results.length > 0) {
-          // set the scrollID for the next work item to the one we received from the update
-          workItem.scrollID = scrollID;
-          await maybeQueueQueryCmrWorkItem(tx, workItem, logger);
-        } else {
-          // Failed to create the next work items - fail the job rather than leaving it orphaned
-          // in the running state
-          logger.error('The work item update should have contained results to queue a next work item, but it did not.');
-          const message = 'Harmony internal failure: could not create the next work items for the request.';
-          await completeJob(tx, job, JobStatus.FAILED, logger, message);
-        }
-      } else {
-        // Finished with the chain for this granule
-        if (status != WorkItemStatus.FAILED) {
-          await addJobLinksForFinishedWorkItem(tx, job, results, logger);
+      const continueProcessing = await handleFailedWorkItems(tx, job, workItem, thisStep, status, logger, errorMessage);
+      if (continueProcessing) {
+        const nextWorkflowStep = await getWorkflowStepByJobIdStepIndex(
+          tx, workItem.jobID, workItem.workflowStepIndex + 1,
+        );
+        if (nextWorkflowStep && (status !== WorkItemStatus.FAILED || nextWorkflowStep?.isBatched)) {
+          createdWorkItem = await createNextWorkItems(
+            tx,
+            nextWorkflowStep,
+            logger,
+            workItem,
+            status,
+            allWorkItemsForStepComplete,
+            results,
+            outputItemSizes,
+          );
         }
-        // If all granules are finished mark the job as finished
-        job.completeBatch(thisStep.workItemCount);
-        if (allWorkItemsForStepComplete) {
-          const finalStatus = await getFinalStatusForJob(tx, job);
-          await completeJob(tx, job, finalStatus, logger);
-        } else {
-          // Special case to pause the job as soon as any single granule completes when in the previewing state
-          if (job.status === JobStatus.PREVIEWING) {
-            job.pause();
+
+        if (nextWorkflowStep && status === WorkItemStatus.SUCCESSFUL) {
```

Comment:
the logic in this section is getting kind of confusing for me

Reply:
Yeah, this whole file and `aggregation-batching`. At least we have significant coverage in the tests to capture many of the possible permutations. It should also allow us to refactor things in the future. I considered trying to rewrite to capture all the possible permutations upfront rather than having deeply nested checks, but that would have had a lot of duplicated code. I think the requirements and logic are really complex at this point so it's going to be hard to not have complex branching code.